components/skin/ui/src/main/resources/PhenoTips/Skin.xml

```diff
@@ -507,39 +507,144 @@ document.observe('xwiki:suggest:containerCreated', function(event) {
+      autosave : {
+        enabled: true,
+        frequency : 10 /*seconds */,
+        message: 'Automatic save'
+      }
+    },
+    initialize : function() {
+      this.form = $('inline');
+      if (this.form) {
+        this.isFormDirty = this.isFormDirty.bind(this);
+        this.form._formState = this;
+        var _this = this;
+        var recordFormData = function() {
+          _this.formData = _this.form.serialize();
+        };
+        // Some form elements are still being modified while "xwiki:dom:loaded" is fired.
+        // If the initial state of the form is recorded immediately, there will likely be
+        // differences with the state of the form when navigating away even if the user
+        // hasn't actually changed any data.
+        // Setting a timeout, even zero, delays the initialization until other
+        // form elements have been modified already.
+        setTimeout(recordFormData, 0);
+        // Record the state of the form again once the form is saved.
+        document.observe('xwiki:document:saved', recordFormData);
+        $$("input[name^='action_']").invoke('observe', 'click', recordFormData);
+        // Upon navigating away, check if anything has changed in the form and warn the user
+        // if it did.
+        window.addEventListener('beforeunload', function(event) {
+          if (_this.isFormDirty()) {
+            var message = "$!escapetool.javascript($services.localization.render('phenotips.navigation.beforeunload.warning'))";
+            (event || window.event).returnValue = message;
+            return message;
+          }
+          return undefined;
+        });
+
+        // Set up the smart autosave
+        if (this.config.autosave.enabled) {
+          this.timer = new PeriodicalExecuter(this.saveIfFormDirty.bind(this), this.config.autosave.frequency);
+        }
+      }
+    },
+    isFormDirty : function() {
+      return this.form ? (this.formData != this.form.serialize()) : false;
+    },
+    saveForm : function(successCallback, failureCallback) {
+      document.fire("xwiki:actions:save", {"continue": true, "form" : this.form});
+      // If there's a callback for there's no unsaved data:
+      if (typeof successCallback == 'function') {
+        // Call the callback function
+        document.observe("xwiki:document:saved", function onSaved() {
+          document.stopObserving("xwiki:document:saved", onSaved);
+          successCallback();
+        });
+        // At the same time, if there's NO callback for when the saving failed, ask user if it's OK to proceed anyway with the successCallback
+        if (typeof failureCallback != 'function') {
+          document.observe("xwiki:document:saveFailed", function onSaveFailed() {
+            document.stopObserving("xwiki:document:saveFailed", onSaveFailed);
+            (new XWiki.widgets.ConfirmationBox({
+              onYes: successCallback,
+              onNo: function(){}
```

Comment:
You can use [`Prototype.K`](http://api.prototypejs.org/Prototype/K/) instead of defining a new function.